include/xtensor/xstrided_view.hpp

```diff
@@ -50,23 +50,89 @@ namespace xt
+            //    const typename std::decay_t<CT>::container_type&,
+            //    typename std::decay_t<CT>::container_type&>;
+            using type = decltype(std::declval<CT>().data());
+        };
+
+        template <typename CT>
+        struct flat_storage_type<CT, typename std::enable_if_t<!has_raw_data_interface<std::decay_t<CT>>::value>>
+        {
+            using type = flat_expression_adaptor<CT>;
+        };
+
+        // with raw_data_interface
+        template <class E, std::enable_if_t<has_raw_data_interface<std::decay_t<E>>::value>* = nullptr>
+        inline decltype(auto) get_flat_storage(E& e)
+        {
+            return e.data();
+        }
+
+        template <class E, std::enable_if_t<has_raw_data_interface<std::decay_t<E>>::value>* = nullptr>
+        inline std::size_t get_offset(E&& e)
+        {
+            return e.raw_data_offset();
+        }
+
+        template <class E, std::enable_if_t<has_raw_data_interface<std::decay_t<E>>::value>* = nullptr>
+        inline decltype(auto) get_strides(E&& e)
+        {
+            return e.strides();
+        }
+
+        // without raw_data_interface
+        template <class E, std::enable_if_t<!has_raw_data_interface<std::decay_t<E>>::value>* = nullptr>
+        inline auto get_flat_storage(E& e) -> flat_expression_adaptor<E>
+        {
+            return flat_expression_adaptor<E>(e);
+        }
+
+        template <class E, std::enable_if_t<!has_raw_data_interface<std::decay_t<E>>::value>* = nullptr>
+        inline std::size_t get_offset(E&& /*e*/)
+        {
+            return std::size_t(0);
+        }
 
-    template <class CT, class S, class CD>
-    struct xiterable_inner_types<xstrided_view<CT, S, CD>>
+        template <class E, std::enable_if_t<!has_raw_data_interface<std::decay_t<E>>::value>* = nullptr>
+        inline auto get_strides(E&& e)
+        {
+            dynamic_shape<std::size_t> strides;
+            strides.resize(e.shape().size());
+            compute_strides(e.shape(), DEFAULT_LAYOUT, strides);
+            return strides;
+        }
+    }
+
+    template <class CT, class S>
+    struct xiterable_inner_types<xstrided_view<CT, S>>
     {
         using inner_shape_type = S;
         using inner_strides_type = inner_shape_type;
         using inner_backstrides_type_type = inner_shape_type;
 
         using const_stepper = std::conditional_t<
                 detail::is_indexed_stepper<typename std::decay_t<CT>::stepper>::value,
-                xindexed_stepper<xstrided_view<CT, S, CD>>,
-                xstepper<const xstrided_view<CT, S, CD>>
+                xindexed_stepper<const xstrided_view<CT, S>>,
+                xstepper<const xstrided_view<CT, S>>
             >;
 
-        using stepper = xstepper<xstrided_view<CT, S, CD>>;
+        using stepper = std::conditional_t<
+                detail::is_indexed_stepper<typename std::decay_t<CT>::stepper>::value,
+                xindexed_stepper<xstrided_view<CT, S>>,
```

Comment:
Should be `xindexed_stepper<xstrided_view<CT, S>, false>`

Reply:
... or perhaps 
```c++
xindexed_stepper<xstrided_view<CT, S>, std::is_const<typename xstrided_view<CT, S>::value_type>::value>
```
to enforce a const stepper when the value_type is const.

Reply:
@ukoethe this cannot be a const stepper, just like const iterator is not like const_iterator.

Reply:
What I mean is: if the value_type of the container/view is `const T`, iterator and const_iterator should be identical. Sorry for the confusion.